src/util.js

```diff
@@ -588,3 +588,120 @@ export function dateIsEmpty(date) {
   }
   return true
 }
+
+// OpenHistoricalMap functions
+
+/**
+ * Returns a `Date` object representing the given UTC date components.
+ *
+ * @param year A one-based year in the proleptic Gregorian calendar.
+ * @param month A zero-based month.
+ * @param day A one-based day.
+ * @returns A date object.
+ */
+function dateFromUTC(year, month, day) {
+  const date = new Date(Date.UTC(year, month, day))
+  // Date.UTC() treats a two-digit year as an offset from 1900.
+  date.setUTCFullYear(year)
+  return date
+}
+
+/**
+ * Converts the given ISO 8601-1 date to a decimal year.
+ *
+ * @param isoDate A date string in ISO 8601-1 format.
+ * @returns A floating point number of years since year 0.
+ */
+function decimalYearFromISODate(isoDate) {
+  // Require a valid YYYY, YYYY-MM, or YYYY-MM-DD date, but allow the year
+  // to be a variable number of digits or negative, unlike ISO 8601-1.
+  if (!isoDate || !/^-?\d{1,4}(?:-\d\d){0,2}$/.test(isoDate)) return undefined
+
+  const ymd = isoDate.split('-')
+  // A negative year results in an extra element at the beginning.
+  if (ymd[0] === '') {
+    ymd.shift()
+    ymd[0] *= -1
+  }
+  const year = +ymd[0]
+  const date = dateFromUTC(year, +ymd[1] - 1, +ymd[2])
+  if (Number.isNaN(date)) return undefined
+
+  // Add the year and the fraction of the date between two New Year’s Days.
+  const nextNewYear = dateFromUTC(year + 1, 0, 1).getTime()
+  const lastNewYear = dateFromUTC(year, 0, 1).getTime()
+  return year + (date.getTime() - lastNewYear) / (nextNewYear - lastNewYear)
+}
+
+/**
+ * Returns a modified version of the given filter that only evaluates to
+ * true if the feature coincides with the given decimal year.
+ *
+ * @param filter The original layer filter.
+ * @param decimalYear The decimal year to filter by.
+ * @returns A filter similar to the given filter, but with added conditions
+ *	that require the feature to coincide with the decimal year.
+ */
+function constrainFilterByDate(filter, decimalYear) {
+  const newFilter = filter
+  if (filter && filter[0] === 'all' && filter[1] && filter[1][0] === 'any') {
+    if (
+      filter[1][2] &&
+      filter[1][2][0] === '<=' &&
+      filter[1][2][1] === 'start_decdate'
+    ) {
+      newFilter[1][2][2] = decimalYear
+    }
+    if (
+      newFilter[2][2] &&
+      newFilter[2][2][0] === '>=' &&
+      newFilter[2][2][1] === 'end_decdate'
+    ) {
+      newFilter[2][2][2] = decimalYear
+    }
+    return newFilter
+  }
+
+  const dateFilter = [
+    'all',
+    ['any', ['!has', 'start_decdate'], ['<=', 'start_decdate', decimalYear]],
```

Comment:
This filter is currently designed to show only the features that existed as of the very moment that is passed in. Since we’re passing in a whole number, it’s only going to show what existed as of January 1 of the year. Would that be alright, or should we loosen the filter to show anything that existed at any time within the given year? The downside is that sometimes you’ll get overlapping features, like two different shops in the same place (because one replaced the other partway through the year) or a construction area on top of the finished construction (because construction finished partway through the year).

```suggestion
      filter[1][2][0] === '<' &&
      filter[1][2][1] === 'start_decdate'
    ) {
      newFilter[1][2][2] = decimalYear + 1
    }
    if (
      newFilter[2][2] &&
      newFilter[2][2][0] === '>=' &&
      newFilter[2][2][1] === 'end_decdate'
    ) {
      newFilter[2][2][2] = decimalYear
    }
    return newFilter
  }

  const dateFilter = [
    'all',
    ['any', ['!has', 'start_decdate'], ['<', 'start_decdate', decimalYear + 1]],
```

Reply:
I think showing anything that existed during the year makes indeed more sense in our case.

The problem with overlapping construction site & building would happen regardless, just shifted by one year, isn't it?

Reply:
The filter is currently written as a precise moment in time. We would show two overlapping features if they’re only tagged with year precision, or if they’re tagged with day precision but just happen to change over on New Year’s Day. It’s probably not a big deal, come to think of it. Someone who cares about this imprecision would want the ability to set a specific date anyways.